File: dbtools/getter/getter.go
package getter

import (
	"fmt"

	"../dbhelp"
	"github.com/mikespook/gearman-go/worker"
	"gopkg.in/mgo.v2/bson"
)

func Create() error {
	// Create the worker and connect it to Gearman

	dbhelp.Log(0, "Creating Gearman worker 'db-get'")
	w := worker.New(worker.OneByOne)
	w.ErrorHandler = func(e error) {
		fmt.Println(e)
	}
	w.AddServer("tcp4", "127.0.0.1:4730")
	w.AddFunc("db-get", DBGet, worker.Unlimited)
	if err := w.Ready(); err != nil {
		dbhelp.Log(2, "Fatal error: "+err.Error())
		return err
	}
	w.Work()
	return nil
}

type DBData struct {
	// the layout of a supplied document
	Key        string `bson:"key"`
	Database   string `bson:"database"`
	Collection string `bson:"collection"`
	Query      bson.M `bson:"query"`
	Projection bson.M `bson:"projection"`
}

func GetDocuments(dbUrl, database, collection string, query bson.M, projection bson.M) ([]byte, error) {
	dbhelp.Log(0, "Getting documents from db "+database+" collection "+collection)
	coll := dbhelp.CreateSession(dbUrl, database, collection)
	var results []bson.M
	err := coll.Find(query).Select(projection).All(&results)
	if err != nil {
		dbhelp.Log(2, "Could not iterate documents: "+err.Error())
		b, _ := bson.Marshal(bson.M{"status": "error", "error": err.Error()})
		return b, err
	}
	response := bson.M{"status": "ok", "docs": results}
	var bsonResponse []byte
	bsonResponse, err = bson.Marshal(response)
	if err != nil {
		dbhelp.Log(2, "Could not marshal bson: "+err.Error())
		b, _ := bson.Marshal(bson.M{"status": "error", "description": err.Error()})
		return b, err
	}
	coll.Database.Session.Close()
	return bsonResponse, err
}

func DBGet(job worker.Job) ([]byte, error) {
	// a gearman wrapper for GetDocuments
	dbhelp.Log(0, "Database getter called!")
	var data DBData
	if err := bson.Unmarshal(job.Data(), &data); err != nil {
		dbhelp.Log(2, err.Error())
		return nil, err
	}
    if !dbhelp.CorrectKey(data.Key) {
        dbhelp.Log(2, "Secret key mismatch")
        b, _ := bson.Marshal(bson.M{"status": "error", "description": "Secret key mismatch"})
        return b, job.Err()
    }
	url := dbhelp.GetURL()
	response, err := GetDocuments(url, data.Database, data.Collection, data.Query, data.Projection)
	if err != nil {
		dbhelp.Log(1, err.Error())
		job.SendWarning([]byte("\"error\":\"" + err.Error() + "\""))
	}
	err = job.Err()
	return response, err
}

File: dbtools/getter/getter.py
import gearman
import pymongo
import bson # the pymongo bson lib

print "Starting python worker 'db-get'"
print "Initiating gearman worker"
gearman_worker = gearman.GearmanWorker(["localhost:4730"])
print "Initiating mongo client"
mongo_client = pymongo.MongoClient('localhost', 27017)

def task_listener_db_getter(worker, job):
    print "Worker initiated: "
    if not bson.is_valid(job.data):
        print "Error: invalid bson"
        return str(bson.BSON.encode({"status":"error", "error":"invalid bson"}))

    try:
        bsonObj = bson.BSON(job.data)
        #data = bson.BSON.decode(bsonObj)
        data = bsonObj.decode()
        print data
        if not (("query" in data) and ("database" in data) and ("collection" in data)):
            print "Missing a field in the query document: " + str(data)
            return str(bson.BSON.encode({"status":"error", "error":"Missing field in query document; one of 'query', 'database' or 'collection'."}))
        db = data["database"]
        coll = data["collection"]
        query = data["query"]
        projection = {}
        if "projection" in data:
            projection = data["projection"]
        print "getting db"
        docs = mongo_client[db][coll].find(query, projection)
        resultDocs = []
        index = 0
        for doc in docs:
            resultDocs.append(doc)
        results = {"docs":resultDocs, "status":"ok"}
        print results
        resultBson = bson.BSON.encode(results)
        return str(resultBson)
    except Exception as e:
        print e
        result = {"status":"error", "error":str(e)}
        resultBson = bson.BSON.encode(result)
        return str(resultBson)

print "Registering worker"
gearman_worker.set_client_id("python-db-getter")
print "Registering task 'db-get'"
gearman_worker.register_task("db-get", task_listener_db_getter)

gearman_worker.work()
